### PR TITLE
[booking] The matching function and quality check tools

### PR DESCRIPTION
**Pull request by syershov**
MAPSME-1281

### CODE REVIEW

File: generator/booking_dataset.cpp [hunk 1]
```diff
@@ -257,11 +257,20 @@ bool BookingDataset::MatchWithBooking(OsmElement const & e) const
     return false;
 
   // Find 3 nearest values to a point.
-  auto const indexes = GetNearestHotels(e.lat, e.lon, 3, 150 /* max distance in meters */);
-  if (indexes.empty())
+  auto const bookingIndexes = GetNearestHotels(e.lat, e.lon, 3, kDistanceLimitInMeters);
```
Comment (ygorshenin) [inline generator/booking_dataset.cpp]: Add comment to the magic 3 constant, please.

File: generator/booking_dataset.cpp [hunk 2]
```diff
@@ -257,11 +257,20 @@ bool BookingDataset::MatchWithBooking(OsmElement const & e) const
...
+  if (bookingIndexes.empty())
```
Comment (ygorshenin) [inline generator/booking_dataset.cpp]: Remove this if, please.

File: generator/booking_dataset.cpp [hunk 3]
```diff
@@ -257,11 +257,20 @@ bool BookingDataset::MatchWithBooking(OsmElement const & e) const
...
     return false;
 
-  bool matched = MatchByName(name, indexes);
+  bool matched = false;
+
+  for (size_t const j : bookingIndexes)
+  {
+    auto const & hotel = GetHotel(j);
+    double const dist = ms::DistanceOnEarth(e.lat, e.lon, hotel.lat, hotel.lon);
```
Comment (ygorshenin) [inline generator/booking_dataset.cpp]: s/dist/distanceMeters/g

File: generator/booking_dataset.hpp [hunk 4]
```diff
@@ -15,6 +15,9 @@ namespace generator
 class BookingDataset
 {
 public:
+  double static constexpr kDistanceLimitInMeters = 150;
+  double static constexpr kOptimalThreshold = 0.709283;
```
Comment (ygorshenin) [inline generator/booking_dataset.hpp]: Comment about how this magic constant was calculated (and how the reader may calculate it).

File: generator/booking_quality_check/booking_quality_check.cpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
+#include "std/iostream.hpp"
+
+#include "generator/booking_dataset.hpp"
+#include "generator/osm_source.hpp"
+
+#include "geometry/distance_on_sphere.hpp"
+
+#include "3party/gflags/src/gflags/gflags.h"
+
+#include "std/fstream.hpp"
```
Comment (ygorshenin) [inline generator/booking_quality_check/booking_quality_check.cpp]: Please, reorder includes - group std and put 3party at the end.

File: generator/booking_quality_check/booking_quality_check.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+      "Takes OSM XML data from stdin and creates"
+      " data and index files in several passes.");
+  google::ParseCommandLineFlags(&argc, &argv, true);
+
+  LOG_SHORT(LINFO, ("Booking data:", FLAGS_booking_data));
+
+  BookingDataset bookingDataset(FLAGS_booking_data);
+
+  vector<OsmElement> elements;
+  auto const counterAction = [&](OsmElement * e)
```
Comment (ygorshenin) [inline generator/booking_quality_check/booking_quality_check.cpp]: Please, get rid of this variable - just write lambda at the call place.

File: generator/booking_quality_check/booking_quality_check.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+      elements.emplace_back(*e);
+  };
+
+  LOG_SHORT(LINFO, ("OSM data:", FLAGS_osm_file_name));
+  {
+    SourceReader reader =
+        FLAGS_osm_file_name.empty() ? SourceReader() : SourceReader(FLAGS_osm_file_name);
+    ProcessOsmElementsFromO5M(reader, counterAction);
+  }
+  LOG_SHORT(LINFO, ("Tourism elements:", elements.size()));
```
Comment (ygorshenin) [inline generator/booking_quality_check/booking_quality_check.cpp]: Num of tourism elements

File: generator/booking_quality_check/booking_quality_check.cpp [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+  vector<size_t> elementIndexes(elements.size());
```
Comment (ygorshenin) [inline generator/booking_quality_check/booking_quality_check.cpp]:
```c++
vector<size_t> elementIndexes(elements.size());
for (size_t i = 0; i < elementIndexes.size(); ++i)
  elementIndexes[i] = i;
```

File: generator/booking_quality_check/booking_quality_check.cpp [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+  size_t counter = 0;
+  for (auto & e : elementIndexes)
+    e = counter++;
+
+  random_shuffle(elementIndexes.begin(), elementIndexes.end());
+  elementIndexes.resize(FLAGS_selection_size);
```
Comment (ygorshenin) [inline generator/booking_quality_check/booking_quality_check.cpp]:
```c++
if (FLAGS_selection_size < elementIndexes.size())
  elementIndexes.resize(FLAGS_selection_size);
```

File: generator/booking_quality_check/booking_quality_check.cpp [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+  stringstream outStream;
+
+  for (size_t i : elementIndexes)
+  {
+    OsmElement const & e = elements[i];
+    auto const bookingIndexes =
+        bookingDataset.GetNearestHotels(e.lat, e.lon, 3, BookingDataset::kDistanceLimitInMeters);
```
Comment (ygorshenin) [inline generator/booking_quality_check/booking_quality_check.cpp]: I see this constant second time. Time to put it into header.

File: generator/booking_quality_check/booking_quality_check.cpp [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    for (size_t const j : bookingIndexes)
+    {
+      auto const & hotel = bookingDataset.GetHotel(j);
+      double const dist = ms::DistanceOnEarth(e.lat, e.lon, hotel.lat, hotel.lon);
+      double score =
```
Comment (ygorshenin) [inline generator/booking_quality_check/booking_quality_check.cpp]: I see this code second time. Time to put it into header.

File: generator/booking_quality_check/booking_quality_check.cpp (hunk 8 above)
Comment (mpimenov) [inline generator/booking_quality_check/booking_quality_check.cpp]:
```
vector<size_t> elementIndexes(elements.size());
iota(elementIndexes.begin(), elementIndexes.end());
```

File: generator/booking_quality_check/booking_quality_check.cpp [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+  {
+    ofstream file(FLAGS_sample_data);
+    if (file.is_open())
+      file << outStream.str();
+    else
+      LOG(LERROR, ("Can't output into", FLAGS_sample_data));
+  }
+
+  return 0;
+}
```
Comment (ygorshenin) [inline generator/booking_quality_check/booking_quality_check.cpp]: newline, please.

File: generator/booking_quality_check/booking_quality_check.cpp (hunk 8 above)
Comment (syershov) [inline generator/booking_quality_check/booking_quality_check.cpp]: done

File: generator/booking_quality_check/booking_quality_check.cpp (hunk 7 above)
Comment (syershov) [inline generator/booking_quality_check/booking_quality_check.cpp]: done

File: generator/booking_quality_check/booking_quality_check.cpp [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+  for (size_t i : elementIndexes)
+  {
+    OsmElement const & e = elements[i];
+    auto const bookingIndexes =
+        bookingDataset.GetNearestHotels(e.lat, e.lon, 3, BookingDataset::kDistanceLimitInMeters);
+    for (size_t const j : bookingIndexes)
+    {
+      auto const & hotel = bookingDataset.GetHotel(j);
+      double const dist = ms::DistanceOnEarth(e.lat, e.lon, hotel.lat, hotel.lon);
```
Comment (ygorshenin) [inline generator/booking_quality_check/booking_quality_check.cpp]: s/dist/distanceMeters/

File: generator/booking_quality_check/booking_quality_check.cpp [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    ProcessOsmElementsFromO5M(reader, counterAction);
+  }
+  LOG_SHORT(LINFO, ("Tourism elements:", elements.size()));
+
+  vector<size_t> elementIndexes(elements.size());
+  size_t counter = 0;
+  for (auto & e : elementIndexes)
+    e = counter++;
+
+  random_shuffle(elementIndexes.begin(), elementIndexes.end());
```
Comment (ygorshenin) [inline generator/booking_quality_check/booking_quality_check.cpp]: random_shuffle is deprecated. Also, please, add a flag for the seed, and use it here.

File: generator/booking_quality_check/booking_quality_check.cpp (hunk 6 above)
Comment (syershov) [inline generator/booking_quality_check/booking_quality_check.cpp]: done

File: generator/booking_quality_check/booking_quality_check.cpp (hunk 5 above)
Comment (syershov) [inline generator/booking_quality_check/booking_quality_check.cpp]: done

File: generator/booking_dataset.cpp (hunk 3 above)
Comment (syershov) [inline generator/booking_dataset.cpp]: done

File: generator/booking_dataset.cpp (hunk 2 above)
Comment (syershov) [inline generator/booking_dataset.cpp]: done

File: generator/booking_dataset.cpp (hunk 1 above)
Comment (syershov) [inline generator/booking_dataset.cpp]: See the line above.

File: generator/booking_dataset.hpp (hunk 4 above)
Comment (syershov) [inline generator/booking_dataset.hpp]: done

File: generator/booking_quality_check/booking_quality_check.cpp (hunk 10 above)
Comment (syershov) [inline generator/booking_quality_check/booking_quality_check.cpp]: done

File: generator/booking_quality_check/booking_quality_check.cpp (hunk 11 above)
Comment (syershov) [inline generator/booking_quality_check/booking_quality_check.cpp]: done

File: generator/booking_quality_check/booking_quality_check.cpp (hunk 12 above)
Comment (syershov) [inline generator/booking_quality_check/booking_quality_check.cpp]: done

File: generator/booking_quality_check/booking_quality_check.cpp (hunk 13 above)
Comment (syershov) [inline generator/booking_quality_check/booking_quality_check.cpp]: done

File: generator/booking_quality_check/booking_quality_check.cpp (hunk 14 above)
Comment (syershov) [inline generator/booking_quality_check/booking_quality_check.cpp]: random_suffle ignores srand seed

File: generator/booking_quality_check/booking_quality_check.cpp (hunk 9 above)
Comment (syershov) [inline generator/booking_quality_check/booking_quality_check.cpp]: done

**Comment by syershov** [comment]
@ygorshenin @mpimenov PTAL

File: generator/booking_dataset.cpp [hunk 15]
```diff
@@ -257,11 +262,18 @@ bool BookingDataset::MatchWithBooking(OsmElement const & e) const
...
+  auto const bookingIndexes = GetNearestHotels(e.lat, e.lon, kMaxSelectedElements, kDistanceLimitInMeters);
+
+  bool matched = false;
+
+  for (size_t const j : bookingIndexes)
+  {
+    auto const & hotel = GetHotel(j);
+    double const distanceMeters = ms::DistanceOnEarth(e.lat, e.lon, hotel.lat, hotel.lon);
+    double score = ScoreByLinearNormDistance(distanceMeters);
+    matched = score > kOptimalThreshold;
```
Comment (ygorshenin) [inline generator/booking_dataset.cpp]: matched = matched || score > kOptimalThreshold

File: generator/booking_quality_check/booking_quality_check.cpp [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+                                if (bookingDataset.TourismFilter(*e))
+                                  elements.emplace_back(*e);
+                              });
+  }
+  LOG_SHORT(LINFO, ("Num of tourism elements:", elements.size()));
+
+  vector<size_t> elementIndexes(elements.size());
+  iota(elementIndexes.begin(), elementIndexes.end(), 0);
+
+  random_shuffle(elementIndexes.begin(), elementIndexes.end());
```
Comment (ygorshenin) [inline generator/booking_quality_check/booking_quality_check.cpp]: Please, leave a comment here why do we need to use random_shuffle here.

File: tools/python/booking_hotels_quality.py [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+def load_binary_list(path):
+    bits = []
+    with open(path, 'r') as fd:
+        for line in fd:
+            if (not line.strip()) or line[0] == '#':
+                continue
+            bits.append(1 if line[0] == 'y' else 0)
+    return bits
+
+def load_score_list(path):
```
Comment (ygorshenin) [inline tools/python/booking_hotels_quality.py]: Comment to the function - what it does.

File: tools/python/booking_hotels_quality.py [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+    with open(path, 'r') as fd:
+        for line in fd:
+            if (not line.strip()) or line[0] == '#':
+                continue
+            scores.append(float(line[line.rfind(':')+2:]))
+    return scores
+
+def process_options():
+    parser = argparse.ArgumentParser(description='Download and process booking hotels.')
+    parser.add_argument("-v", "--verbose", action="store_true", dest="verbose")
```
Comment (ygorshenin) [inline tools/python/booking_hotels_quality.py]: Please, use either single or double quotes.

File: tools/python/booking_hotels_quality.py [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+    return options
+
+def main():
+    options = process_options()
+    reference = load_binary_list(options.reference_list)
+    sample = load_score_list(options.sample_list)
+
+    precision, recall, threshold = metrics.precision_recall_curve(reference, sample)
+    aa = zip(precision, recall, threshold)
+    print("Optimal thrashold: {2} for precision: {0} and recall: {1}".format(*max(aa, key=lambda (p, r, t): p*r/(p+r))))
```
Comment (ygorshenin) [inline tools/python/booking_hotels_quality.py]: threshold

File: tools/python/booking_hotels_quality.py [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    print("AUC: {0}".format(metrics.roc_auc_score(reference, sample)))
+
+    if options.show:
+        import matplotlib.pyplot as plt
```
Comment (ygorshenin) [inline tools/python/booking_hotels_quality.py]: Move this to the top of the file.

File: tools/python/booking_hotels_quality.py [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+        plt.plot(recall, precision)
```
Comment (ygorshenin) [inline tools/python/booking_hotels_quality.py]: Please, add a title, xlab, ylab.

File: generator/booking_quality_check/booking_quality_check.cpp [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+  }
+
+  if (FLAGS_sample_data.empty())
+  {
+    cout << outStream.str();
+  }
+  else
+  {
+    ofstream file(FLAGS_sample_data);
+    if (file.is_open())
```
Comment (mpimenov) [inline generator/booking_quality_check/booking_quality_check.cpp]: Может быть, эту проверку стоит вынести в начало?

File: generator/booking_dataset.cpp (hunk 15 above)
Comment (syershov) [inline generator/booking_dataset.cpp]: done

File: generator/booking_quality_check/booking_quality_check.cpp (hunk 16 above)
Comment (syershov) [inline generator/booking_quality_check/booking_quality_check.cpp]: done

File: tools/python/booking_hotels_quality.py (hunk 17 above)
Comment (syershov) [inline tools/python/booking_hotels_quality.py]: done

File: tools/python/booking_hotels_quality.py (hunk 18 above)
Comment (syershov) [inline tools/python/booking_hotels_quality.py]: done

File: tools/python/booking_hotels_quality.py (hunk 20 above)
Comment (syershov) [inline tools/python/booking_hotels_quality.py]: done

File: tools/python/booking_hotels_quality.py (hunk 21 above)
Comment (syershov) [inline tools/python/booking_hotels_quality.py]: done

File: generator/booking_quality_check/booking_quality_check.cpp (hunk 22 above)
Comment (syershov) [inline generator/booking_quality_check/booking_quality_check.cpp]: Так файл создается строчкой выше, наверное все же лучше оставить.

**Comment by syershov** [comment]
@ygorshenin @mpimenov PTAL

File: generator/booking_dataset.cpp [hunk 23]
```diff
@@ -257,11 +262,18 @@ bool BookingDataset::MatchWithBooking(OsmElement const & e) const
...
+  auto const bookingIndexes = GetNearestHotels(e.lat, e.lon, kMaxSelectedElements, kDistanceLimitInMeters);
+
+  bool matched = false;
+
+  for (size_t const j : bookingIndexes)
+  {
+    auto const & hotel = GetHotel(j);
+    double const distanceMeters = ms::DistanceOnEarth(e.lat, e.lon, hotel.lat, hotel.lon);
+    double score = ScoreByLinearNormDistance(distanceMeters);
+    matched = matched || score > kOptimalThreshold;
```
Comment (mpimenov) [inline generator/booking_dataset.cpp]:
```
  for (...)
  {
  ...
  if (score > kOptimalThreshold)
    return true;
  }
return fasle;
```

File: generator/booking_dataset.cpp [hunk 24]
```diff
@@ -223,6 +223,11 @@ void BookingDataset::BuildFeatures(function<void(OsmElement *)> const & fn) cons
   }
 }
 
+double BookingDataset::ScoreByLinearNormDistance(double distance) const
+{
+  return (kDistanceLimitInMeters - distance) / kDistanceLimitInMeters;
```
Comment (ygorshenin) [inline generator/booking_dataset.cpp]:
```c++
distance = clamp(distance, 0, kDistanceLimitInMeters);
return 1.0 - distance / kDistanceLimitInMeters;
```

File: generator/booking_dataset.hpp [hunk 25]
```diff
@@ -63,6 +69,8 @@ class BookingDataset
 
   void BuildFeatures(function<void(OsmElement *)> const & fn) const;
 
+  double ScoreByLinearNormDistance(double distance) const;
```
Comment (ygorshenin) [inline generator/booking_dataset.hpp]: static

File: tools/python/booking_hotels_quality.py [hunk 26]
```diff
@@ -0,0 +1,87 @@
...
+import time
+import urllib2
+
+# init logging
+logging.basicConfig(level=logging.DEBUG, format='[%(asctime)s] %(levelname)s: %(message)s')
+
+
+def load_binary_list(path):
+    """
+    Loads binary classifier output
```
Comment (ygorshenin) [inline tools/python/booking_hotels_quality.py]: Dot at the end.

File: tools/python/booking_hotels_quality.py [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+        for line in fd:
+            if (not line.strip()) or line[0] == '#':
+                continue
+            bits.append(1 if line[0] == 'y' else 0)
+    return bits
+
+
+def load_score_list(path):
+    """
+    Loads list of scores
```
Comment (ygorshenin) [inline tools/python/booking_hotels_quality.py]: Dot at the end.

Comment (ygorshenin) [inline tools/python/booking_hotels_quality.py]: s/scores/matching scores/g

File: generator/booking_quality_check/booking_quality_check.cpp [hunk 28]
```diff
@@ -0,0 +1,108 @@
...
+                                  elements.emplace_back(*e);
+                              });
+  }
+  LOG_SHORT(LINFO, ("Num of tourism elements:", elements.size()));
+
+  vector<size_t> elementIndexes(elements.size());
+  iota(elementIndexes.begin(), elementIndexes.end(), 0);
+
+  // In first implementation, we used random_shufle for reference dataset.
+  // Next time we are going to replace random_shuffle by shuffle with defined seed.
```
Comment (mpimenov) [inline generator/booking_quality_check/booking_quality_check.cpp]: todo?

File: tools/python/booking_hotels_quality.py [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+
+
+def main():
+    options = process_options()
+    reference = load_binary_list(options.reference_list)
+    sample = load_score_list(options.sample_list)
+
+    precision, recall, threshold = metrics.precision_recall_curve(reference, sample)
+    aa = zip(precision, recall, threshold)
+    print("Optimal threshold: {2} for precision: {0} and recall: {1}".format(*max(aa, key=lambda (p, r, t): p*r/(p+r))))
```
Comment (ygorshenin) [inline tools/python/booking_hotels_quality.py]:
```python
2 * p * r / (p + r)
```

File: generator/osm_source.cpp [hunk 30]
```diff
@@ -545,6 +550,15 @@ bool GenerateFeaturesImpl(feature::GenerateInfo & info)
...
       LOG(LINFO, ("Processing booking data from", info.m_bookingDatafileName, "done."));
+      string skippedElementsPath = info.GetIntermediateFileName("skipped_elements", ".lst");
+      ofstream file(skippedElementsPath);
+      if (file.is_open())
+      {
+        file << skippedElements.str();
+        LOG(LINFO, ("Saving skipped elements if into", skippedElementsPath, "done."));
+      }
+      else
+        LOG(LERROR, ("Can't output into", skippedElementsPath));
```
Comment (mpimenov) [inline generator/osm_source.cpp]: {
}

File: generator/osm_source.cpp [hunk 31]
```diff
@@ -545,6 +550,15 @@ bool GenerateFeaturesImpl(feature::GenerateInfo & info)
     {
       bookingDataset.BuildFeatures([&](OsmElement * e) { parser.EmitElement(e); });
       LOG(LINFO, ("Processing booking data from", info.m_bookingDatafileName, "done."));
+      string skippedElementsPath = info.GetIntermediateFileName("skipped_elements", ".lst");
+      ofstream file(skippedElementsPath);
+      if (file.is_open())
+      {
+        file << skippedElements.str();
+        LOG(LINFO, ("Saving skipped elements if into", skippedElementsPath, "done."));
```
Comment (mpimenov) [inline generator/osm_source.cpp]: s/if into/into/

File: tools/python/booking_hotels_quality.py (hunk 29 above)
Comment (ygorshenin) [inline tools/python/booking_hotels_quality.py]: Or it's better to:
```python
def hmean(*xs):
    return len(xs) / sum(1 / x for x in xs)
```
and use it here.

Comment (ygorshenin) [inline tools/python/booking_hotels_quality.py]: Or just import from scipy.

Comment (mpimenov) [inline tools/python/booking_hotels_quality.py]: "for" лишнее?

File: generator/booking_dataset.cpp (hunk 23 above)
Comment (syershov) [inline generator/booking_dataset.cpp]: done

File: generator/booking_dataset.cpp (hunk 24 above)
Comment (syershov) [inline generator/booking_dataset.cpp]: done

File: generator/booking_dataset.hpp (hunk 25 above)
Comment (syershov) [inline generator/booking_dataset.hpp]: done

File: generator/booking_quality_check/booking_quality_check.cpp (hunk 28 above)
Comment (syershov) [inline generator/booking_quality_check/booking_quality_check.cpp]: nope

File: generator/osm_source.cpp (hunk 31 above)
Comment (syershov) [inline generator/osm_source.cpp]: done

File: generator/osm_source.cpp (hunk 30 above)
Comment (syershov) [inline generator/osm_source.cpp]: done

File: tools/python/booking_hotels_quality.py (hunk 26 above)
Comment (syershov) [inline tools/python/booking_hotels_quality.py]: done

File: tools/python/booking_hotels_quality.py (hunk 27 above)
Comment (syershov) [inline tools/python/booking_hotels_quality.py]: done

**Comment by syershov** [comment]
@ygorshenin @mpimenov PTAL

**Comment by ygorshenin** [comment]
LGTM

**Comment by mpimenov** [comment]
LGTM

**Comment by syershov** [comment]
Thanx for review!

**Pull request merged**